sdk-utils/src/main/kotlin/com/rakuten/tech/mobile/sdkutils/network/NetworkUtil.kt

```diff
@@ -0,0 +1,105 @@
+
+    constructor(context: Context): this(context, null)
+
+    /**
+     * Initializes this class by registration to network callback.
+     * Call this API before invoking other APIs.
+     *
+     * @throws Exception exception when net capabilities retrieval fails on versions on or before
+     * Android 8.
+     */
+    @Throws(Exception::class)
+    fun initialize() {
+        if (appContext.hasPermission(Manifest.permission.ACCESS_NETWORK_STATE)) {
+            try {
+                registerNetworkCallback()
+            } catch (ex: Exception) {
+                logger.debug("Network callback registration failed", ex)
+                isNetworkAvailable.set(false)
+                netCapabilities.set(null)
+                throw ex
+            }
+        }
+    }
+
+    /**
+     * Returns whether the device has network connectivity.
+     *
+     * @return true if online, false otherwise.
+     */
+    fun isOnline() = isNetworkAvailable.get()
+
+    /**
+     * Returns the capabilities of the active network.
+     *
+     * @return NetworkCapabilities object
+     */
+    fun networkCapabilities() = netCapabilities.get()
+
+    private fun registerNetworkCallback() {
+        val request = NetworkRequest.Builder()
+            .addTransportType(NetworkCapabilities.TRANSPORT_WIFI)
+            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
+            .build()
+        val connectivityManager =
+            appContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        connectivityManager.registerNetworkCallback(
+            request,
+            object : ConnectivityManager.NetworkCallback() {
+                override fun onCapabilitiesChanged(
+                    network: Network,
+                    networkCapabilities: NetworkCapabilities
+                ) {
+                    super.onCapabilitiesChanged(network, networkCapabilities)
+                    netCapabilities.set(networkCapabilities)
+                }
+
+                override fun onLost(network: Network) {
+                    super.onLost(network)
+                    isNetworkAvailable.set(false)
+                    netCapabilities.set(null)
+                }
+
+                override fun onAvailable(network: Network) {
+                    super.onAvailable(network)
+                    isNetworkAvailable.set(true)
+
+                    // workaround: onCapabilitiesChanged is not guaranteed to be called immediately after
+                    // onAvailable in versions below Android 8
+                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+                        try {
+                            netCapabilities.set(connectivityManager.getNetworkCapabilities(network))
+                        } catch (ex: Exception) {
+                            logger.debug("Network capabilities retrieval failed", ex)
+                            netCapabilities.set(null)
```

Comment:
Should the `isNetworkAvailable` be set to `false`  here?

Reply:
No since the device is still online, only that the network capabilities cannot be retrieved on this instance.
Please see the following for the referred implementation: analytics-rat > DeviceUtil.kt